### PR TITLE
Move self trait predicate to items

### PR DESCRIPTION
**Pull request by nikomatsakis**
This is a "reimagination" of @tmandry's PR #50183. The main effect is described in this comment from one of the commits:

---

Before we had the following results for `predicates_of`:

```rust
trait Foo { // predicates_of: Self: Foo
  fn bar(); // predicates_of: Self: Foo (inherited from trait)
}
```

Now we have removed the `Self: Foo` from the trait. However, we still
add it to the trait ITEM. This is because when people do things like
`<T as Foo>::bar()`, they still need to prove that `T: Foo`, and
having it in the `predicates_of` seems to be the cleanest way to
ensure that happens right now (otherwise, we'd need special case code
in various places):

```rust
trait Foo { // predicates_of: []
  fn bar(); // predicates_of: Self: Foo
}
```

However, we sometimes want to get the list of *just* the predicates
truly defined on a trait item (e.g., for chalk, but also for a few
other bits of code). For that, we define `predicates_defined_on`,
which does not contain the `Self: Foo` predicate yet, and we plumb
that through metadata and so forth.

---

I'm assigning @eddyb as the main reviewer, but I thought I might delegate to scalexm for this one in any case. I also want to post an alternative that I'll leave in the comments; it occurred to me as I was writing. =)

r? @eddyb 
cc @scalexm @tmandry @leodasvacas

### CODE REVIEW

**Comment by nikomatsakis** [comment]
It occurs to me that a less disruptive way to achieve the **main goal** of this PR would be to 

- create `predicates_defined_on` vs `predicates_of`
- continue to add the `Self: Trait` predicate **but for traits**, not trait items

(We would have to rework slightly how the `instantiate` mechanism works, since we would want to instantiate the `predicates_defined_on` of the parent -- I realize now that this is always instantiating `predicates_of` of the parent, but that's ok at the moment.)

Crucially, this would still mean that the rest of the compiler thinks that `Foo: Trait` being WF requires `Foo: Trait` to hold. This would also mean we don't have to modify `ParamEnv`. Those things mean less "churn" -- but they are not necessarily **good**. After all, the WF requirement is just overly zealous.

For background, I think that the way that @scalexm and I **wanted** things to work is that we add the `Self: Trait` assumption in the `ParamEnv` **only** (and not in `predicates_of`). And then things that project out of traits also prove that those traits hold (e.g., method invocation code etc). But it wasn't obvious to me how to do that cleanly -- in some sense, the `predicates_of` and `predicates_defined_on` split was our attempt to do that cleanly.

File: src/librustc/ty/sty.rs [hunk 1]
```diff
@@ -614,6 +614,14 @@ impl<'tcx> TraitRef<'tcx> {
         TraitRef { def_id: def_id, substs: substs }
     }
 
+    /// Returns a TraitRef of the form `Self: Foo<U>`.
+    pub fn identity<'a, 'gcx>(tcx: TyCtxt<'a, 'gcx, 'tcx>, def_id: DefId) -> TraitRef<'tcx> {
+        TraitRef {
+            def_id,
+            substs: Substs::identity_for_item(tcx, def_id),
```
Comment (eddyb) [inline src/librustc/ty/sty.rs]: Can you make sure no existing code uses `identity_for_item` with `TraitRef`?

Comment (nikomatsakis) [reply src/librustc/ty/sty.rs]: yeah

File: src/librustc/ty/mod.rs [hunk 2]
```diff
@@ -2887,6 +2892,13 @@ fn param_env<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
     traits::normalize_param_env_or_error(tcx, def_id, unnormalized_env, cause)
 }
 
+pub fn is_trait<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId) -> bool {
+    match tcx.def_key(def_id).disambiguated_data.data {
+        DefPathData::Trait(_) => true,
+        _ => false,
+    }
+}
```
Comment (eddyb) [inline src/librustc/ty/mod.rs]: Why is this a free function and not a method on `tcx`? Also it's a bit sad to see such a specific hack but I don't have any good suggestion.

File: src/librustc/ty/sty.rs (hunk 1 above)
Comment (nikomatsakis) [reply src/librustc/ty/sty.rs]: found a few uses, commit incoming

File: src/librustc/ty/mod.rs (hunk 2 above)
Comment (nikomatsakis) [reply src/librustc/ty/mod.rs]: Good question, should be a method on tcx I agree. By hack you are referring to the whole param-env change, right? (Not the fn itself)

I'm somewhat torn, as I wrote elsewhere -- in a way, I think adding things to param-env is the *right* behavior -- or maybe the answer is that `predices_of` on traits should include `Self: Trait` but WF should use `predicates_defined_on`.

**Comment by nikomatsakis** [comment]
Regarding the `param_env` change, I'm pulling out this part of my earlier comment for further thought:

> I'm somewhat torn, as I wrote elsewhere -- in a way, I think adding things to param-env is the right behavior -- or maybe the answer is that predices_of on traits should include Self: Trait but WF should use predicates_defined_on.

**Comment by nikomatsakis** [comment]
OK, so I've been thinking a bit more about this. I didn't fully remember how the `predicates_of` query works -- in particular, it returns the predicates **on the current item** but includes only the def-d of the parent (the list is unrolled in the `instantiate` step).

this has got me wondering. Maybe we should do this:

- Keep the `predicates_defined_on` split
- Add `Self: Trait` to `predicates_of` for the trait
- Have the WF requirements use `predicates_defined_on` and not `predicates_of`

Then we can not change `ParamEnv` (it uses `predicates_of`); we don't have to change the method "comparison" code (though we can) -- it works today because the `Self: Trait` is found in the **parent** predicate list, which is ignored. We don't have to change the object safety code either (though, again, we can), since it too ignores the parent list.

In fact, `predicates_defined_on` could maybe be called `wf_predicates_of` or something -- for everything but traits, this is the same as `predicates_of`.

**Comment by tmandry** [comment]
> Add `Self: Trait` to predicates_of for the trait

Either way sounds fine to me.

Picking the minimally invasive path makes sense. Would this roll back some of the changes you made to tests? I hit some of those, and decided it was best for this change not to modify current behavior (unless it's fixing a clear deficiency.)

Also, it seems we've come full circle 😅

**Comment by nikomatsakis** [comment]
@tmandry it would not affect the tests. We could keep the current behavior by making the "well-formed" rules using `predicates_of`, not `predicates_defined_on`. I'm torn because keeping the behavior seems good but then that behavior *is* also wrong.

**Comment by nikomatsakis** [comment]
> Also, it seems we've come full circle on predicates_of 😅

Yeah =) that design is basically saying "`predicates_of` was right as is, but we needed something else sometimes". =)

**Comment by nikomatsakis** [comment]
I just pushed a new commit with the [approach I outlined here](https://github.com/rust-lang/rust/pull/51895#issuecomment-401336806).

> So with the change, the behavior is still wrong, just manifested differently, correct?

On the branch as I have it now, the behavior is not wrong. That is, proving that `T: Trait` is well-formed does not require proving that `T: Trait` is true. However...

> I would only be worried about other unintended consequences we don't currently test for. Admittedly I don't understand the cause of the wrong behavior in the first place, which is why I'm apprehensive....I'm also assuming that these WF checks will be subsumed by chalk at some point.

Yes, this is the source of my apprehension as well. We've certainly hit surprises when touching things like this. It might be wise to try to leave the behavior of rustc unaffected and then let chalk fix things in a more principled way, no question.

I'm not sure what consequences this change might have. We could try a crater run to see. I'm reminded though of some of the complications we encountered after https://github.com/rust-lang/rust/pull/48557 landed. 

OK, here is a proposal: I will make this PR preserve behavior (but help with Chalk) and we can land it.  Then I will open a separate PR that fixes the WF behavior and we can do a crater run.

**Comment by nikomatsakis** [comment]
OK, the most recent commit reverts rustc's "WF" behavior so that proving that `T: Foo` is well-formed also requires proving `T: Foo`. This definitely seems wrong to me but it avoids affecting any tests -- the remaining tests that are affected are from a drive-by rewrite of the code around defaulted bounds, which I'd prefer to keep, and anyway it only affects the details of the error message, not which errors you get.

**Comment by nikomatsakis** [comment]
I'll cleanup the commit history again

**Comment by nikomatsakis** [comment]
Cleaned up the history now.

File: src/librustc_typeck/collect.rs [hunk 3]
```diff
@@ -1324,9 +1324,35 @@ fn predicates_of<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
...
                            def_id: DefId)
                            -> ty::GenericPredicates<'tcx> {
+    let ty::GenericPredicates { parent, mut predicates } =
+        tcx.predicates_defined_on(def_id);
+
+    if tcx.is_trait(def_id) {
+        // For traits, add `Self: Trait` predicate. This is
+        // not part of the predicates that a user writes, but it
+        // is something that one must prove in order to invoke a
+        // method method or project an associated type.
```
Comment (scalexm) [inline src/librustc_typeck/collect.rs]: Duplicate word « method »

**Review by scalexm — APPROVED** [review]

**Comment by nikomatsakis** [comment]
cc @rust-lang/infra -- I'm assuming these errors reported above are spurious:

```
[00:04:12]       Memory: 8 GB
[00:04:12]       Boot ROM Version: VMW71.00V.0.B64.1704110547
[00:04:12]       Apple ROM Info: [MS_VM_CERT/SHA1/27d66596a61c48dd3dc7216fd715126e33f59ae7]Welcome to the Virtual Machine
[00:04:12]       SMC Version (system): 2.8f0
[00:04:12]       Serial Number (system): VMzcdHi9sZvn
[00:04:12] 
[00:04:12] hw.ncpu: 4
[00:04:12] hw.byteorder: 1234
[00:04:12] hw.memsize: 8589934592
---
$ rvm $(travis_internal_ruby) --fuzzy do ruby -S gem install dpl
Warning, new version of rvm available '1.29.4', you are using older version '1.29.3'.
You can disable this warning with:    echo rvm_autoupdate_flag=0 >> ~/.rvmrc
You can enable  auto-update  with:    echo rvm_autoupdate_flag=2 >> ~/.rvmrc
ERROR:  While executing gem ... (Gem::RemoteFetcher::FetchError)
    Errno::EHOSTUNREACH: Failed to open TCP connection to api.rubygems.org:443 (No route to host - connect(2) for "api.rubygems.org" port 443) (https://api.rubygems.org/quick/Marshal.4.8/dpl-1.9.7.gemspec.rz)

```

**Pull request merged**